### PR TITLE
implement macros

### PR DESCRIPTION
**Pull request by invisig0th**

### CODE REVIEW

File: synapse/lib/stormlib/macro.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        user = self.runt.user
+        if mdef['user'] != user.iden and not user.isAdmin():
+            mesg = 'Macro belongs to a different user'
+            raise s_exc.AuthDeny(mesg=mesg)
+
+        await self.runt.snap.core.popHiveKey(path)
+
+    async def _funcMacroSet(self, name, storm):
+
+        name = await s_stormtypes.toprim(name)
```
Comment (vEpiphyte) [inline synapse/lib/stormlib/macro.py:151]: We need to `isinstance` check that the name is a string.

File: synapse/lib/stormtypes.py [hunk 2]
```diff
@@ -915,6 +919,12 @@ def __init__(self, valu, path=None):
     def __len__(self):
         return len(self.valu)
 
+    def __bool__(self):
+        return bool(self.valu)
+
+    def __str__(self):
+        return self.valu.decode()
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:926]: This is unsafe and should not be used to str-ify bytes. We have no gaurante that the data held in self.valu is valid utf8 bytes. We'll throw UnicodeDecodeError's into the runtime.

Comment (invisig0th) [reply synapse/lib/stormtypes.py:926]: I think it's safe to assume a default encoding of utf8 ( for explicit conversion they can use the decode() method directly ) for an implicit tostr() cast.  I'll catch Unicode exception and wrap it into a badcast.

File: synapse/cortex.py [hunk 3]
```diff
@@ -1892,12 +1894,17 @@ async def _initStormCmds(self):
         self.addStormCmd(s_storm.SpliceListCmd)
         self.addStormCmd(s_storm.SpliceUndoCmd)
 
+        self.addStormCmd(s_stormlib_macro.MacroExecCmd)
```
Comment (vEpiphyte) [inline synapse/cortex.py:1897]: extra whitespace

File: synapse/cortex.py [hunk 4]
```diff
@@ -1935,6 +1942,7 @@ def _initStormLibs(self):
         self.addStormLib(('user',), s_stormtypes.LibUser)
         self.addStormLib(('vars',), s_stormtypes.LibVars)
         self.addStormLib(('view',), s_stormtypes.LibView)
+        self.addStormLib(('macro',), s_stormlib_macro.LibMacro)
```
Comment (vEpiphyte) [inline synapse/cortex.py:1945]: pyramid

File: synapse/lib/stormlib/macro.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        if not self.runtsafe:
+            mesg = 'macro.exec does not support per-node invocation'
+            raise s_exc.StormRuntimeError(mesg=mesg)
+
+        name = await s_stormtypes.tostr(self.opts.name)
+        path = ('cortex', 'storm', 'macros', name)
+
+        mdef = await runt.snap.core.getHiveKey(path)
+        if mdef is None:
+            # FIXME Bad ref "name"
```
Comment (vEpiphyte) [inline synapse/lib/stormlib/macro.py:104]: remove fixme

File: synapse/lib/stormtypes.py [hunk 6]
```diff
@@ -1705,8 +1724,22 @@ def __init__(self, runt, ldef, path=None):
             'get': self._methLayerGet,
             'pack': self._methLayerPack,
             'repr': self._methLayerRepr,
+            'edits': self._methLayerEdits,
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:1727]: This is not safe for httpapi callers to use

File: synapse/lib/stormtypes.py [hunk 7]
```diff
@@ -2915,10 +2948,22 @@ async def toprim(valu, path=None):
         return valu
 
     if isinstance(valu, (tuple, list)):
-        return tuple([await toprim(v) for v in valu])
+        retn = []
+        for v in valu:
+            try:
+                retn.append(await toprim(v))
+            except s_exc.NoSuchType:
+                pass
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:2956]: silently dropping objects and changing the size of an list/tuple feels scary.

File: synapse/lib/stormtypes.py [hunk 8]
```diff
@@ -2915,10 +2948,22 @@ async def toprim(valu, path=None):
...
+        return tuple(retn)
 
     if isinstance(valu, dict):
-        return {await toprim(k): await toprim(v) for (k, v) in valu.items()}
+        retn = {}
+        for k, v in valu.items():
+            try:
+                retn[k] = await toprim(v)
+            except s_exc.NoSuchType:
+                pass
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:2965]: silently dropping objects and changing the size of a dict feels scary.

File: synapse/lib/stormlib/macro.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        self.locls.update({
+            'set': self._funcMacroSet,
+            'get': self._funcMacroGet,
+            'del': self._funcMacroDel,
+            'list': self._funcMacroList,
+        })
+
+    async def _funcMacroList(self):
+        path = ('cortex', 'storm', 'macros')
+        return await self.runt.snap.core.getHiveKeys(path)
```
Comment (vEpiphyte) [inline synapse/lib/stormlib/macro.py:136]: Attempting to use this with with spawn ends up with an attribute error:
`ERROR: ('AttributeError', {'efile': 'macro.py', 'eline': 136, 'mesg': "'SpawnCore' object has no attribute 'getHiveKeys'"})`

File: synapse/lib/stormlib/macro.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        storm = await s_stormtypes.tostr(storm)
+
+        # validation
+        await self.runt.getStormQuery(storm)
+
+        path = ('cortex', 'storm', 'macros', name)
+
+        user = self.runt.user
+
+        mdef = await self.runt.snap.core.getHiveKey(path)
```
Comment (vEpiphyte) [inline synapse/lib/stormlib/macro.py:172]: attempting to use this with spawn fails

File: synapse/lib/stormlib/macro.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+        name = await s_stormtypes.tostr(name)
+
+        path = ('cortex', 'storm', 'macros', name)
+        return await self.runt.snap.core.getHiveKey(path)
+
+    async def _funcMacroDel(self, name):
+        name = await s_stormtypes.tostr(name)
+        path = ('cortex', 'storm', 'macros', name)
+
+        mdef = await self.runt.snap.core.getHiveKey(path)
```
Comment (vEpiphyte) [inline synapse/lib/stormlib/macro.py:148]: attempting to use this with spawn fails

File: synapse/lib/stormlib/macro.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+
+    async def _funcMacroList(self):
+        path = ('cortex', 'storm', 'macros')
+        return await self.runt.snap.core.getHiveKeys(path)
+
+    async def _funcMacroGet(self, name):
+        name = await s_stormtypes.tostr(name)
+
+        path = ('cortex', 'storm', 'macros', name)
+        return await self.runt.snap.core.getHiveKey(path)
```
Comment (vEpiphyte) [inline synapse/lib/stormlib/macro.py:142]: attempting to use this with spawn fails

File: synapse/lib/stormlib/macro.py [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+                $lib.print('{name} (owner: {user})', name=$name.ljust(20), user=$user.name)
+                $count = $($count + 1)
+            }
+            $lib.print('{count} macros found', count=$count)
+        ''',
+    },
+]
+
+class MacroExecCmd(s_storm.Cmd):
+    # FIXME Give a description of use.
```
Comment (jnwatson) [inline synapse/lib/stormlib/macro.py:81]: ahem fixme

File: synapse/lib/stormlib/macro.py (hunk 5 above)
Comment (jnwatson) [inline synapse/lib/stormlib/macro.py:104]: fixme

File: synapse/lib/stormlib/macro.py [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+
+    def getArgParser(self):
+        pars = s_storm.Cmd.getArgParser(self)
+        pars.add_argument('name', help='The name of the macro to execute')
+        return pars
+
+    async def execStormCmd(self, runt, genr):
+
+        if not self.runtsafe:
+            mesg = 'macro.exec does not support per-node invocation'
```
Comment (jnwatson) [inline synapse/lib/stormlib/macro.py:96]: really? This seems a major limitation

File: synapse/lib/stormlib/macro.py [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+        name = await s_stormtypes.tostr(name)
+        path = ('cortex', 'storm', 'macros', name)
+
+        mdef = await self.runt.snap.core.getHiveKey(path)
+        if mdef is None:
+            mesg = f'Macro name not found: {name}'
+            raise s_exc.NoSuchName(mesg)
+
+        user = self.runt.user
+        if mdef['user'] != user.iden and not user.isAdmin():
```
Comment (jnwatson) [inline synapse/lib/stormlib/macro.py:154]: Let's discuss perms

File: synapse/lib/parser.py [hunk 16]
```diff
@@ -395,6 +395,7 @@ def massage_vartokn(x):
     'DOUBLEQUOTEDSTRING': lambda x: s_ast.Const(unescape(x)),  # drop quotes and handle escape characters
     'NUMBER': lambda x: s_ast.Const(s_ast.parseNumber(x)),
     'HEXNUMBER': lambda x: s_ast.Const(s_ast.parseNumber(x)),
+    'BOOL': lambda x: s_ast.Bool(x == 'true'),
```
Comment (jnwatson) [inline synapse/lib/parser.py:398]: this is weird

File: synapse/lib/stormlib/macro.py (hunk 14 above)
Comment (vEpiphyte) [reply synapse/lib/stormlib/macro.py:96]: doing this in a non-runtsafe fashion means we have to get the macro defintion for each node inline, potentially.

this prevents doing 
```
macro.exec $macroName
```

File: synapse/cortex.py (hunk 4 above)
Comment (invisig0th) [reply synapse/cortex.py:1945]: depends on which element we're pyramiding on :)

File: synapse/lib/parser.py (hunk 16 above)
Comment (invisig0th) [reply synapse/lib/parser.py:398]: The const value of a bool is limited to the string "true" and the string "false" so this works perfectly :)

File: synapse/lib/stormtypes.py (hunk 6 above)
Comment (invisig0th) [reply synapse/lib/stormtypes.py:1727]: yep, ack.

**Pull request merged**